casetokens.php

```diff
@@ -144,6 +144,46 @@ function _casetokens_get_case_id() {
   return isset(\Civi::$statics['casetokens']['case_id']) ? \Civi::$statics['casetokens']['case_id'] : NULL;
 }
 
+/**
+ * Get all the contact entity fields.
+ *
+ * @return array
+ */
+function _casetokens_get_contact_fields() {
+  $contactId = CRM_Core_Session::singleton()->getLoggedInContactID();
+  $contactFields = array();
+  try {
+    $contactFields = civicrm_api3('contact', 'getsingle', array(
+      'id' => $contactId,
+    ));
+  } catch (Throwable $ex) {
+  }
+
+  return array_keys($contactFields);
```

Comment:
This is not the correct way to fetch field metadata. You should use `civicrm_api3('contact', 'getfields')`

Reply:
I understand and initially I opted for that api but actually this getfields returns a lot of unrelated fields for example api_key, hash,contact_sub_type to name a few so if I opt for this I will have add extra computation to to remove all those unrequired fields

Reply:
`get` will also return things like api_key and hash if the logged in user has them. And that's the main problem with your approach here - the api doesn't return NULL values, so if the current logged-in user has a NULL `nick_name` for example, then it won't come back as a field. Too unpredictable.
I think the best option is to use getfields and remove items you don't want.

Reply:
Api method has been changed as per the suggestions